// drink/src/main/webapp/drive/js/app/require_config.js
( function() {
  require.config( {
    deps: [ 'initialize_app' ],
	paths : {
		'backbone' : '../../../components/backbone/backbone',
		'hammerjs' : '../../../components/hammerjs/hammer',
		'jquery-hammerjs' : '../../../components/jquery-hammerjs/jquery.hammer',
		'backbone-hammer' : '../../../components/backbone.hammer/backbone.hammer',
		'underscore' : '../../../components/lodash/dist/lodash',
		'jquery' : '../../../components/jquery/dist/jquery',
		'jquery-validation' : '../../../components/jquery.validation/dist/jquery.validate.min',
		'handlebars' : '../../../components/handlebars/handlebars',
		'hbs' : '../../../components/hbs/hbs',
		'lib' : 'jquery-lib',
		'starRating' : 'jquery.rating.js',
		//'d3': '../../../components/d3/d3',
		'd3' : '../../../components/d3/d3.min',
		'i18next' : '../../../components/i18next/i18next.amd',
		//'i18n': '../../../components/i18next/i18next.amd.min',
		//'handlebars-i18next': '../../../components/handlebars-i18next/handlebars-i18next',
		'handlebars-i18next' : 'handlebars-i18next',
		'hbsStarRating' : 'hbsStarRating',
		//'handlebars-i18next': '../../../components/handlebars-i18next/handlebars-i18next.min'
		'moment' : '../../../components/moment/min/moment-with-locales',
		'moment-isoduration' : '../../../components/moment-isoduration/moment.isoduration',
		'validator' : '../../../components/validator.js/dist/validator',
		'bootstrap' : '../../../components/bootstrap/dist/js/bootstrap',
		//'transitions': '../../../components/backbone.responsiveCSS3transitions/backbone.responsiveCSS3transitions',
		'page-transitions' : 'page-transitions',
		'jknob' : '../../../components/jknob/js/jquery.knob',
		'jquery-easing-original' : '../../../components/jquery-easing-original/jquery.easing.1.3',
		'raygun' : '../../../components/raygun4js/dist/raygun',
		'fastclick' : '../../../components/fastclick-amd/fastclick'
	},
		hbs: {
				templateExtension: 'html',
				helperDirectory: 'hbs.helpers/',
				helpers: true,
				i18n: false,
				helperPathCallback: function ( name ) {
					var plugin;
					switch( name ) {
						
						case 'i18next':
						case 'i18nextr':
							plugin = 'handlebars-i18next';
							break;
					
						case 'hbsStarRating':
							plugin = 'hbsStarRating';
							break;
							
						default:
							plugin = 'templates/helpers/' + name;
							break;
					}
					
					return plugin;
				}
		},
    shim: {
      backbone: {
        deps: [ 'jquery', 'underscore' ],
        exports: 'Backbone'
      },
			'jquery-validation': {
        deps: [ 'jquery' ]
      },
			'jknob': {
        deps: [ 'jquery' ]
      },
			
			'jquery-easing-original': {
        deps: [ 'jquery' ]
      },
			'lib': {
        deps: [ 'jquery' ]
      },
			'starRating': {
        deps: [ 'jquery' ]
      },

			'moment-isoduration': {
				deps: [ 'moment' ]
			},

			'bootstrap': {
        deps: [ 'jquery' ]
      }

    }
  } );

} ).call( this );
